test/router.ts

```diff
@@ -0,0 +1,317 @@
+
+    it("doesn't allow extra keys", () => {
+      const confTmpl = (v) => `
+routes:
+  wrong:
+    matchers:
+      title: ["test"]
+    output:
+      type: "alert"${v}
+      dimensions: ["dim1", "dim2"]
+      stat_type: "gauge"
+`;
+
+      const actual = new router.Router();
+      const validConf = confTmpl(`
+      series: "whatever"`);
+      actual._loadConfigString(validConf);
+
+      const invalidConf = confTmpl(`
+      series: "whatever"
+      something-else: "hi there"`);
+      assert.throws(() => actual._loadConfigString(invalidConf));
+    });
+  });
+
+  describe("route", () => {
+    it("works", () => {
+      const r = new router.Router([
+        new router.Rule(
+          "rule-one",
+          {title: ["hello", "hi"], foo: ["bar", "baz"]},
+          {channel: "#-%{foo}-", dimensions: ["-%{foo}-"]}
+        ),
+        new router.Rule(
+          "rule-two",
+          {"bing.bong": ["buzz"]},
+          {series: "x"}
+        ),
+      ]);
+
+      const msg0 = {
+        title: "hi",
+        foo:   "bar",
+      };
+      const expected0 = [{
+        rule:       "rule-one",
+        channel:    "#-bar-",
+        dimensions: ["-bar-"],
+      }];
+      const actual0 = r.route(msg0).routes;
+      assert.deepEqual(expected0, actual0);
+
+      const msg1 = {
+        title: "hi",
+        bing: {
+          bong: "buzz",
+        },
+      };
+      const expected1 = [{
+        rule:   "rule-two",
+        series: "x",
+      }];
+      const actual1 = r.route(msg1).routes;
+      assert.deepEqual(expected1, actual1);
+
+      const msg2 = {
+        title: "hello",
+        foo:   "baz",
+        bing: {
+          bong: "buzz",
+        },
+      };
+      const expected2 = [
+        {
+          rule:       "rule-one",
+          channel:    "#-baz-",
+          dimensions: ["-baz-"],
+        },
+        {
+          rule:   "rule-two",
+          series: "x",
+        },
+      ];
+      const actual2 = r.route(msg2).routes;
+      assert.deepEqual(expected2, actual2);
+    });
+  });
+});
+
+describe("router.Rule", () => {
+  describe("matches", () => {
+    it("works on simple cases", () => {
+      const r = new router.Rule("test-rule", {title: ["hello", "hi"], foo: ["bar"]}, {});
+      const msg0 = {
+        title: "hello",
+        foo:   "bar",
+      };
+      const msg1 = {
+        title: "hi",
+        foo:   "bar",
+      };
+      const msg2 = {
+        title: "hi",
+        foo:   "fighters",
+      };
+      const msg3 = {
+        title: "howdy",
+        foo:   "bar",
+      };
+      const msg4 = {
+        "missing-stuff": "indeed",
+      };
+      assert(r.matches(msg0));
+      assert(r.matches(msg1));
+      assert(!r.matches(msg2));
+      assert(!r.matches(msg3));
+      assert(!r.matches(msg4));
+    });
+
+    it("works on nested messages", () => {
+      const r = new router.Rule("test-rule", {"foo.bar": ["hello", "hi"]}, {});
+      const msg0 = {
+        title: "greeting",
```

Comment:
It should be irrelevant but I decided to leave it in just to make the tests more realistic.